src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+ *  but all other windows (if any) will receive events.
+ *
+ *  @a stdin_text will be gfree()-d automatically when it's sent entirely to the child, or
+ *  when the child exits.
+ *
+ *  The text passed to @a stdout_cb and @a stderr_cb is read with @c g_io_channel_read_line(),
+ *  but is guaranteed to be non-@c NULL.
+ *  After @a exit_cb is invoked, there will be no more data to read.
+ *
+ *  @param working_directory child's current working directory, or @c NULL.
+ *  @param command_line child program and arguments, or @c NULL.
+ *  @param argv child's argument vector, or @c NULL.
+ *  @param envp child's environment, or @c NULL.
+ *  @param block_window window whose message loop to block until the child exits, or @c NULL.
+ *  @param stdin_text text to send to childs's stdin, or @c NULL.
+ *  @param stdout_cb callback to receive child's stdout, or @c NULL.
+ *  @param stdout_data data to pass to @a stdout_cb.
+ *  @param stderr_cb callback to receive child's stderr, or @c NULL.
+ *  @param stdout_data data to pass to @a stderr_cb, or NULL.
+ *  @param exit_cb callback to invoke when the child exits, or @c NULL.
+ *  @param exit_data data to pass to @a exit_cb.
+ *  @param child_pid return location for child process ID, or @c NULL.
+ *  @param error return location for error.
+ *
+ *  @return @c TRUE on success, @c FALSE if an error was set.
+ **/
+gboolean spawn_with_callbacks(const gchar *working_directory, const gchar *command_line,
+	gchar **argv, gchar **envp, GtkWindow *block_window, gchar *stdin_text,
+	void (*stdout_cb)(const gchar *text, gpointer user_data), gpointer stdout_data,
+	void (*stderr_cb)(const gchar *text, gpointer user_data), gpointer stderr_data,
+	void (*exit_cb)(gint status, gpointer user_data), gpointer exit_data, GPid *child_pid,
+	GError **error)
+{
+	GSource *source = g_source_new(&spawn_source_funcs, sizeof(SpawnSource));
+	SpawnSource *sp = (SpawnSource *) source;
+
+	sp->std[IN].fd = sp->std[OUT].fd = sp->std[ERR].fd = -1;
+	sp->std[IN].events = G_IO_OUT | G_IO_ERR;
+	sp->std[OUT].events = sp->std[ERR].events = G_IO_IN | G_IO_HUP | G_IO_ERR;
+
+	if (spawn_async_with_pipes(working_directory, command_line, argv, envp, &sp->pid,
+		stdin_text ? &sp->std[IN].fd : NULL, stdout_cb ? &sp->std[OUT].fd : NULL,
+		stderr_cb ? &sp->std[ERR].fd : NULL, error))
+	{
+		int i;
+
+		if (child_pid)
+			*child_pid = sp->pid;
+
+		if (stdin_text)
+		{
+		#ifdef G_OS_WIN32
+			sp->last_send_ticks = GetTickCount();
+		#else
+			g_source_add_poll(source, &sp->std[IN]);
+		#endif
+			sp->stdin_pos = sp->stdin_text = stdin_text;
+			sp->stdin_remaining = strlen(sp->stdin_text);
+		}
+
+		sp->io_cb[OUT] = stdout_cb;
+		sp->io_data[OUT] = stdout_data;
+		sp->io_cb[ERR] = stderr_cb;
+		sp->io_data[ERR] = stderr_data;
+
+		for (i = 0; i < 2; i++)
+		{
+			if (sp->std[i].fd != -1)
+			{
+			#ifdef G_OS_WIN32
+				sp->ioc[i] = g_io_channel_win32_new_fd(sp->std[i].fd);

```

Comment:
It doesn't work for named pipes. IIRC, glib doesn't even try to implement it, for example because you can't check child's stdin. Works for sockets, and for handles that support async I/O - but creates thead(s?) for the latter.